### PR TITLE
Typecheck user-defined functions

### PR DESCRIPTION
**Pull request by AndreaRinaldi1**
In this branch I implement the early typecheck for arguments and ultimately for return type for user-defined functions. 

Just like for builtin functions, TypePromotionIterator wraps every argument that is not null and that has a SequenceType type definition in the function signature different from Item*. 
Analogously, it also wraps the whole function to check on its return type if different from Item*. 

As you can see from the tests, this works both for static user-defined function calls and dynamic function calls.

### CODE REVIEW

File: src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/DynamicFunctionCallIterator.java [hunk 1]
```diff
@@ -23,6 +23,7 @@
 import org.apache.spark.api.java.JavaRDD;
 import org.rumbledb.api.Item;
 import sparksoniq.exceptions.IteratorFlowException;
+import sparksoniq.exceptions.SequenceExceptionExactlyOne;
```
Comment (wscsprint3r) [inline src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/DynamicFunctionCallIterator.java]: Is this needed?

File: src/main/java/sparksoniq/jsoniq/runtime/iterator/operational/TypePromotionIterator.java [hunk 2]
```diff
@@ -161,6 +159,8 @@ public boolean initIsRDD() {
     }
 
     private void checkTypePromotion() {
+        if (_nextResult.isFunction())
```
Comment (ghislainfourny) [inline src/main/java/sparksoniq/jsoniq/runtime/iterator/operational/TypePromotionIterator.java]: Good catch. Thanks!

File: src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/base/Functions.java [hunk 3]
```diff
@@ -579,7 +579,20 @@ public static RuntimeIterator buildUserDefinedFunctionCallIterator(
             IteratorMetadata metadata,
             List<RuntimeIterator> arguments
     ) {
-        return new FunctionItemCallIterator(functionItem, arguments, metadata);
+        FunctionItemCallIterator functionCallIterator = new FunctionItemCallIterator(functionItem, arguments, metadata);
```
Comment (ghislainfourny) [inline src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/base/Functions.java]: Perfect!

File: src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/FunctionItemCallIterator.java [hunk 4]
```diff
@@ -92,6 +95,25 @@ private void processArguments() {
             );
         }
 
+        if (_functionItem.getSignature().getParameterTypes() != null) {
+            for (int i = 0; i < _functionArguments.size(); i++) {
+                if (
+                    _functionArguments.get(i) != null
+                        && !_functionItem.getSignature().getParameterTypes().get(i).equals(mostGeneralSequenceType)
+                ) {
+                    _functionArguments.set(
```
Comment (ghislainfourny) [inline src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/FunctionItemCallIterator.java]: Very good and clean.

File: src/main/java/sparksoniq/exceptions/UnknownFunctionCallException.java [hunk 5]
```diff
@@ -39,4 +39,8 @@ public UnknownFunctionCallException(String fnName, int arity, IteratorMetadata i
             iteratorMetadata.getExpressionMetadata()
         );
     }
+
+    public UnknownFunctionCallException(String message, IteratorMetadata iteratorMetadata) {
```
Comment (ghislainfourny) [inline src/main/java/sparksoniq/exceptions/UnknownFunctionCallException.java]: Where is this used? I cannot see any new call in the diff.

File: src/main/java/sparksoniq/semantics/visitor/RuntimeIteratorVisitor.java [hunk 6]
```diff
@@ -430,23 +427,12 @@ public RuntimeIterator visitFunctionDeclaration(FunctionDeclaration expression,
...
         FunctionItem function = new FunctionItem(
-                new FunctionIdentifier(expression.get_name(), parameters.size()),
-                parametersNames,
-                new FunctionSignature(
-                        parameters,
-                        returnType
-                ),
-                bodyIterator,
-                new HashMap<>()
+                expression.get_name(),
```
Comment (ghislainfourny) [inline src/main/java/sparksoniq/semantics/visitor/RuntimeIteratorVisitor.java]: Very nice, indeed this is a code duplication.

File: src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/DynamicFunctionCallIterator.java (hunk 1 above)
Comment (AndreaRinaldi1) [reply src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/DynamicFunctionCallIterator.java]: No, probably just something added by mistake. 
Thank you

File: src/main/java/sparksoniq/exceptions/UnknownFunctionCallException.java (hunk 5 above)
Comment (AndreaRinaldi1) [reply src/main/java/sparksoniq/exceptions/UnknownFunctionCallException.java]: You are right, I probably added that when I added the exception for non-existing function, but then I decided to stick to the existing one. 
Thank you for spotting

**Pull request merged**